### PR TITLE
fix(pocket): Remove Save to Pocket option if Pocket is disabled

### LINKED ISSUE #3268: Remove Save to Pocket option if Pocket is disabled
asa noted that one can disable `extensions.pocket.enabled` and that should hide the option. Clicking the option when disabled results in a blank page info arrow panel.

image

### PR DESCRIPTION
**Pull request by csadilek**
Closes #3268

### CODE REVIEW

File: system-addon/lib/SectionsManager.jsm [hunk 1]
```diff
@@ -39,13 +40,19 @@ const BUILT_IN_SECTIONS = {
...
+  CONTEXT_MENU_PREFS: {"SaveToPocket": "extensions.pocket.enabled"},
   initialized: false,
   sections: new Map(),
   init(prefs = {}) {
     for (const feedPrefName of Object.keys(BUILT_IN_SECTIONS)) {
       const optionsPrefName = `${feedPrefName}.options`;
       this.addBuiltInSection(feedPrefName, prefs[optionsPrefName]);
     }
+
+    this.prefsObserver = () => this.updateSections();
```
Comment (Mardak) [inline system-addon/lib/SectionsManager.jsm]: Just add an `observe` method to the object and then you can call `addObserver(pref, this)`

https://github.com/mozilla/activity-stream/blob/17e5e5bd4bd5917cda4114840feb375d14db2b9f/system-addon/lib/ActivityStream.jsm#L278-L288

File: system-addon/lib/SectionsManager.jsm [hunk 2]
```diff
@@ -77,18 +85,34 @@ const SectionsManager = {
...
+
     if (this.sections.has(id)) {
       this.sections.set(id, Object.assign(this.sections.get(id), options));
       this.emit(this.UPDATE_SECTION, id, options, shouldBroadcast);
     }
   },
+  updateSectionContextMenuOptions(options) {
+    if (options.availableContextMenuOptions) {
+      options.contextMenuOptions = options.availableContextMenuOptions.filter(
+        o => !this.CONTEXT_MENU_PREFS[o] || Services.prefs.getBoolPref(this.CONTEXT_MENU_PREFS[o]));
```
Comment (Mardak) [inline system-addon/lib/SectionsManager.jsm]: Add a comment here or for the whole method that this is to show the option unless there's a pref to conditionally show the option.

File: system-addon/lib/SectionsManager.jsm [hunk 3]
```diff
@@ -77,18 +85,34 @@ const SectionsManager = {
...
   },
+  updateSections() {
+    this.sections.forEach((section, id) => this.updateSection(id, section, true));
+  },
   updateSection(id, options, shouldBroadcast) {
+    this.updateSectionContextMenuOptions(options);
+
     if (this.sections.has(id)) {
       this.sections.set(id, Object.assign(this.sections.get(id), options));
       this.emit(this.UPDATE_SECTION, id, options, shouldBroadcast);
```
Comment (Mardak) [inline system-addon/lib/SectionsManager.jsm]: Do we care if a section might get an "UPDATE" even if it didn't care about the pref for a context menu item?

**Comment by csadilek** [comment]
Thanks @Mardak. Nits fixed.

File: system-addon/lib/SectionsManager.jsm (hunk 3 above)
Comment (csadilek) [reply system-addon/lib/SectionsManager.jsm]: Sounds like a good improvement once we have a second menu item that's affected by this or a second section that doesn't have SaveToPocket (Highlights will also have SaveToPocket).

File: system-addon/lib/SectionsManager.jsm [hunk 4]
```diff
@@ -39,16 +40,32 @@ const BUILT_IN_SECTIONS = {
...
+      Services.prefs.addObserver(this.CONTEXT_MENU_PREFS[k], this));
+
     this.initialized = true;
     this.emit(this.INIT);
   },
+  observe(subject, topic, data) {
+    switch (topic) {
+      case "nsPref:changed":
+        for (const pref of Object.keys(this.CONTEXT_MENU_PREFS)) {
+          if (data === this.CONTEXT_MENU_PREFS[pref]) {
```
Comment (Mardak) [inline system-addon/lib/SectionsManager.jsm]: Ha. If only it were a `Map` could have just done `.has(data)` ;)

**Comment by as-pine-proxy** [comment]
[Treeherder: 0f69e56f1415cc00919a20d044ac254f3b09632c](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=0f69e56f1415cc00919a20d044ac254f3b09632c)

**Pull request merged**